docs/source/hooks.rst

```diff
@@ -0,0 +1,173 @@
+    @Project.operation
+    @Project.operation_hooks.on_success(store_success_to_doc)
+    @Project.operation_hooks.on_exception(store_error_to_doc)
+    @Project.post.isfile("result.txt")
+    def error_on_a_0(job):
+        if job.sp.a == 0:
+            # Have jobs with statepoint 'a' == 0 fail
+            raise ValueError
+
+    if __name__ == '__main__':
+       Project().main()
+
+
+If ``error_on_a_0`` is executed using ``python project.py run --operation error_on_a_0 --filter a 1``, the hook triggered ``on_success`` will run, and ``job.doc.error_on_a_0_success`` will be ``True``.
+
+If ``error_on_a_0`` is executed using ``python project.py run --operation error_on_a_0 --filter a 0``, a ``ValueError`` is raised.
+The hook triggered ``on_exception`` will run, and ``job.doc.error_on_a_0_success`` will be ``False``.
+
+
+.. _project-level-hooks:
+
+Project-Level Hooks
+===================
+
+It may be desirable to install the same hook or set of hooks for all operations in a project.
+In the following example FlowProject, the hook ``track_start_time`` is triggered when each operation starts.
+The hook appends the current time to a list in the job document that is named based on the name of the operation.
+
+ .. code-block:: python
+
+    from flow import FlowProject
+
+    class Project(FlowProject):
+        pass
+
+    @Project.operation
+    @Project.post.true('test_ran')
+    def do_operation(job):
+        job.doc.test_ran = True
+
+    @Project.operation
+    @Project.pre.after(do_operation)
+    @Project.post.false('test_ran')
+    def undo_operation(job):
+        job.doc.test_ran = False
+
+    def track_start_time(operation_name, job):
+        import time
+        current_time = time.strftime('%b %d, %Y at %l:%M:%S %p %Z')
+        doc_key = f'{operation_name}_start_times'
+        times = job.doc.get(doc_key, [])
+        times.append(current_time)
+        job.doc[doc_key] = times
+
+    if __name__ == '__main__':
+        project = Project()
+        project.project_hooks.on_start.append(track_start_time)
+        project.main()
+
+
+A custom set of hooks may also be installed at the project level by a custom ``install_hooks`` method.
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject
+
+    class Project(FlowProject):
+        pass
+
+    @Project.operation
+    @Project.post.true('test_ran')
+    def do_operation(job):
+        job.doc.test_ran = True
+
+    # Define custom hooks class.
+    # This can be done in a separate file and imported into the project.py file.
+    class ProjectHooks:
+
+        def __init__(self, project):
+            self.project = project
+
+        def set_job_doc(self, key):
+            def set_true(operation_name, job):
+                job.doc[f"{operation_name}_{key}"] = True
+            return set_true
+
+        def set_job_doc_with_error(self):
+            def set_false(operation_name, error, job):
+                job.doc[f"{operation_name}_success"] = True
```

Comment:
True should be False here.
```suggestion
                job.doc[f"{operation_name}_success"] = False
```

Reply:
Why does the structure of `set_job_doc_with_error` not accept a key `success` like the other?